IRremoteESP8266.cpp

```diff
@@ -1955,48 +1971,55 @@ bool ICACHE_FLASH_ATTR IRrecv::decodeDaikin(decode_results *results) {
   return true;
 }
 
-// Denon, from https://github.com/z3t0/Arduino-IRremote/blob/master/ir_Denon.cpp
-bool ICACHE_FLASH_ATTR IRrecv::decodeDenon (decode_results *results) {
-	unsigned long data   = 0;  // Somewhere to build our code
-	int offset = 1;  // Skip the Gap reading
+// Decode a Denon message.
+//
+// Args:
+//   results: Ptr to the data to decode and where to store the decode result.
+//   nbits:   Expected nr. of data bits. (Typically DENON_BITS)
+// Returns:
+//   boolean: True if it can decode it, false if it can't.
+//
+// Ref: https://github.com/z3t0/Arduino-IRremote/blob/master/ir_Denon.cpp
+bool ICACHE_FLASH_ATTR IRrecv::decodeDenon(decode_results *results,
+                                           uint16_t nbits,
+                                           bool strict) {
+  // Check we have enough data
+  if (results->rawlen < 2 * nbits + HEADER + FOOTER)
+    return false;
+  if (strict && nbits != DENON_BITS)
+    return false;
 
-	// Check we have the right amount of data
-	if (results->rawlen != 1 + 2 + (2 * DENON_BITS) + 1) {
-	  return false;
-	}
+  uint64_t data = 0;
+  uint16_t offset = OFFSET_START;
 
-	// Check initial Mark+Space match
-	if (!matchMark (results->rawbuf[offset++], DENON_HDR_MARK )) {
-	  return false;
-	}
-	if (!matchSpace(results->rawbuf[offset++], DENON_HDR_SPACE)) {
-	  return false;
-	}
-
-	// Read the bits in
-	for (int i = 0;  i < DENON_BITS;  i++) {
-		// Each bit looks like: DENON_MARK + DENON_SPACE_1 -> 1
-		//                 or : DENON_MARK + DENON_SPACE_0 -> 0
-		if (!matchMark(results->rawbuf[offset++], DENON_BIT_MARK)) {
-		  return false;
-		}
-
-		// IR data is big-endian, so we shuffle it in from the right:
-		if (matchSpace(results->rawbuf[offset], DENON_ONE_SPACE)) {
-		  data = (data << 1) | 1;
-		} else if (matchSpace(results->rawbuf[offset], DENON_ZERO_SPACE)) {
-		  data = (data << 1) | 0;
-		} else {
-		  return false;
-		}
-		offset++;
-	}
-
-	// Success
-  results->bits = DENON_BITS;
-	results->value = data;
-	results->decode_type = DENON;
-	return true;
+  // Header
+  if (!matchMark(results->rawbuf[offset++], DENON_HDR_MARK))
+    return false;
+  if (!matchSpace(results->rawbuf[offset++], DENON_HDR_SPACE))
+    return false;
+  // Data
+  for (int i = 0; i < nbits; i++) {
+    if (!matchMark(results->rawbuf[offset++], DENON_BIT_MARK))
```

Comment:
I'm moderately certain we can't, as it would be a [false positive](https://en.wikipedia.org/wiki/False_positives_and_false_negatives#False_positive_error). While each _data bit_ is a pairing of a 'mark + space' and _yes_, it's true that a mark will always be followed by a space. However the length of the mark could be different to expected, hence the need for a check. i.e. The mark signal might be out of spec.

By way of a simplified example for brevity.
(Key: H=header, D=data, F=Footer; S=space, M=mark)

A normal packet would look like `HM,HS  DM,DS DM,DS ... FM,(till next command)S`

I _think_ you are proposing to only check the first DM, and just assume all the following marks are DMs, correct?

If we did, then `HM,HS  DM,DS UnexpectedM,DS DM,DS ... FM,(till next command)S` would still be recognised as a valid packet, potentially at the expense of it later being checked against some other valid/correct protocol.

Have I explained why well enough, or have I misunderstood what you were suggesting, or am I missing something more obvious (which is always/likely possible :)?